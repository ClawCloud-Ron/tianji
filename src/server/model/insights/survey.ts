import { z } from 'zod';
import { insightsQuerySchema } from '../../utils/schema.js';
import { prisma } from '../_client.js';
import { getDateQuery, printSQL } from '../../utils/prisma.js';
import { Prisma } from '@prisma/client';
import dayjs from 'dayjs';
import { FilterInfoType, FilterInfoValue, getDateArray } from '@tianji/shared';
import { mapValues } from 'lodash-es';
import { env } from '../../utils/env.js';
import { buildCommonFilterQueryOperator } from './shared.js';

export async function insightsSurvey(
  query: z.infer<typeof insightsQuerySchema>,
  context: { timezone: string }
): Promise<
  {
    date: string;
  }[]
> {
  const { time } = query;
  const { startAt, endAt, unit, timezone = context.timezone } = time;

  const sql = buildInsightsSurveySql(query, context);

  if (env.isDev) {
    printSQL(sql);
  }

  const res = await prisma.$queryRaw<{ date: string | null }[]>(sql);

  return getDateArray(
    res.map((item) => {
      return {
        ...mapValues(item, (val) => Number(val)),
        date: String(item.date),
      };
    }),
    startAt,
    endAt,
    unit,
    timezone
  );
}

export function buildInsightsSurveySql(
  query: z.infer<typeof insightsQuerySchema>,
  context: { timezone: string }
): Prisma.Sql {
  const { insightId, time, metrics, filters, groups } = query;
  const { startAt, endAt, unit, timezone = context.timezone } = time;

  const selectQueryArr = metrics.map((item) => {
    if (item.math === 'events') {
      if (item.name === '$all_event') {
        return Prisma.sql`count(1) as "$all_event"`;
      }

      return Prisma.sql`sum(case WHEN "SurveyResult"."payload"->>'${item.name}' IS NOT NULL AND "SurveyResult"."payload"->>'${item.name}' <> '' THEN 1 ELSE 0 END) as ${Prisma.raw(`"${item.name}"`)}`;
    } else if (item.math === 'sessions') {
      if (item.name === '$all_event') {
        return Prisma.sql`count(distinct "sessionId") as "$all_event"`;
      }

      return Prisma.sql`count(distinct case WHEN "SurveyResult"."payload"->>'${item.name}' IS NOT NULL AND "SurveyResult"."payload"->>'${item.name}' <> '' THEN "sessionId" ELSE 0 END) as ${Prisma.raw(`"${item.name}"`)}`;
    }
  });

  let groupSelectQueryArr: Prisma.Sql[] = [];
  if (groups && groups.length > 0) {
    for (const g of groups) {
      if (!g.customGroups) {
        groupSelectQueryArr.push(
          Prisma.sql`"payload" ->> ${g.value} as "%${Prisma.raw(g.value)}"`
        );
      } else if (g.customGroups && g.customGroups.length > 0) {
        for (const cg of g.customGroups) {
          groupSelectQueryArr.push(
            Prisma.sql`${buildFilterQueryOperator(
              g.value,
              'string',
              cg.filterOperator,
              cg.filterValue
            )} as "%${Prisma.raw(`${g.value}|${cg.filterOperator}|${cg.filterValue}`)}"`
          );
        }
      }
    }
  }

  let innerJoinQuery = Prisma.empty;
  if (filters.length > 0) {
    innerJoinQuery = Prisma.sql`WHERE ${Prisma.join(
      filters.map((filter) =>
        buildFilterQueryOperator(
          filter.name,
          filter.type,
          filter.operator,
          filter.value
        )
      ),
      ' AND '
    )}`;
  }

  const whereQueryArr = [
    // survey id
    Prisma.sql`"SurveyResult"."surveyId" = ${insightId}`,

    // date
    Prisma.sql`"SurveyResult"."createdAt" between ${dayjs(startAt).toISOString()}::timestamptz and ${dayjs(endAt).toISOString()}::timestamptz`,
  ];

  const groupByText = Array.from({ length: groupSelectQueryArr.length + 1 })
    .map((_, i) => `${i + 1}`)
    .join(' , ');

  const sql = Prisma.sql`select
        ${getDateQuery('"SurveyResult"."createdAt"', unit, timezone)} date,
        ${Prisma.join([...groupSelectQueryArr, ...selectQueryArr], ' , ')}
      from "SurveyResult"
      ${innerJoinQuery}
      where ${Prisma.join(whereQueryArr, ' AND ')}
      group by ${Prisma.raw(groupByText)}`;

  return sql;
}

function buildFilterQueryOperator(
  name: string,
  type: FilterInfoType,
  operator: string,
  value: FilterInfoValue | null
): Prisma.Sql {
  const valueField = Prisma.sql`("SurveyResult"."payload"->> ${name})${type === 'number' ? '::int' : type === 'boolean' ? '::boolean' : Prisma.empty}`;
  return buildCommonFilterQueryOperator(type, operator, value, valueField);
}
