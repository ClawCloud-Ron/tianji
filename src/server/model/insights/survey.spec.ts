import { describe, expect, test } from 'vitest';
import { buildInsightsSurvey } from './survey.js';
import { unwrapSQL } from '../../utils/prisma.js';

describe('buildInsightsSurvey', () => {
  const insightId = 'cm658i2tqw96upkejldn8rpbs';
  const insightType = 'survey';

  test('groups', () => {
    const sql = buildInsightsSurvey(
      {
        insightId,
        insightType,
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [
          {
            value: 'rating',
          },
        ],
      },
      {
        timezone: 'UTC',
      }
    );

    expect(unwrapSQL(sql)).toMatchSnapshot('sql');
  });

  test('groups with custom bucket', () => {
    const sql = buildInsightsSurvey(
      {
        insightId,
        insightType,
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [
          {
            value: 'rating',
            customGroups: [
              {
                filterOperator: 'in list',
                filterValue: ['1', '2', '3'],
              },
            ],
          },
        ],
      },
      {
        timezone: 'UTC',
      }
    );

    expect(unwrapSQL(sql)).toMatchSnapshot('sql');
  });
});
