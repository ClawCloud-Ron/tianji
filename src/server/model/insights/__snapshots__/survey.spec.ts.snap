// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildInsightsSurvey > groups > sql 1`] = `
"select
        to_char(date_trunc('day', "SurveyResult"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date,
        "payload" ->> 'rating' as "%rating" , count(1) as "$all_event"
      from "SurveyResult"
      
      where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz
      group by 1 , 2"
`;

exports[`buildInsightsSurvey > groups with custom bucket > sql 1`] = `
"select
        to_char(date_trunc('day', "SurveyResult"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date,
        ("SurveyResult"."payload"->> 'rating') IN ('1' , '2' , '3') as "%rating|in list|1,2,3" , count(1) as "$all_event"
      from "SurveyResult"
      
      where "SurveyResult"."surveyId" = 'cm658i2tqw96upkejldn8rpbs' AND "SurveyResult"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz
      group by 1 , 2"
`;
